src/nbl/ui/CWindowXCB.cpp

```diff
@@ -0,0 +1,296 @@
+    auto NET_WM_PING = connection->resolveAtom(m_window.m_NET_WM_PING);
+
+    if(auto event = xcb.pxcb_wait_for_event(connection->getRawConnection())) {
+        auto* eventCallback = m_window.getEventCallback();
+        m_window.m_clipboardManager->process(&m_window, event);
+        switch (event->response_type & ~0x80)
+        {
+            case 0: {
+                xcb_generic_error_t* error = reinterpret_cast<xcb_generic_error_t*>(event);
+                printf("XCB error: %d", error->error_code);
+                break;
+            }
+            case XCB_CONFIGURE_NOTIFY: {
+                xcb_configure_notify_event_t* cne = reinterpret_cast<xcb_configure_notify_event_t*>(event);
+                if(m_window.m_width != cne->width || 
+                    m_window.m_height != cne->height) {
+                    eventCallback->onWindowResized(&m_window, cne->width, cne->height);
+                }
+                if(m_window.m_x != cne->x || 
+                    m_window.m_y != cne->y) {
+                    eventCallback->onWindowMoved(&m_window, cne->x, cne->y);
+                }
+                break;
+            }
+            case XCB_DESTROY_WINDOW: {
+                xcb_destroy_window_request_t* dwr = reinterpret_cast<xcb_destroy_window_request_t*>(event);
+                if(dwr->window == m_window.m_xcbWindow) {
+                    m_quit = true;
+                    eventCallback->onWindowClosed(&m_window);
+                }
+                break;
+            }
+            case XCB_CLIENT_MESSAGE: {
+                xcb_client_message_event_t* cme = reinterpret_cast<xcb_client_message_event_t*>(event);
+                if(cme->data.data32[0] == MW_DELETE_WINDOW) {
+                    xcb.pxcb_unmap_window(m_window.getXcbConnection(), m_window.m_xcbWindow);
+                    xcb.pxcb_destroy_window(m_window.getXcbConnection(), m_window.m_xcbWindow);
+                    xcb.pxcb_flush(m_window.getXcbConnection());
+                    m_window.m_xcbWindow = 0;
+                    m_quit = true; // we need to quit the dispatch thread
+                    eventCallback->onWindowClosed(&m_window);
+                } else if(cme->data.data32[0] == NET_WM_PING && cme->window != connection->primaryScreen()->root) {
+                    xcb_client_message_event_t ev = *cme;
+                    ev.response_type = XCB_CLIENT_MESSAGE;
+                    ev.window = m_window.m_xcbWindow;
+                    ev.type = NET_WM_PING;
+                    xcb.pxcb_send_event(m_window.getXcbConnection(), 0, m_window.m_xcbWindow, XCB_EVENT_MASK_NO_EVENT, reinterpret_cast<const char*>(&ev));
+                    xcb.pxcb_flush(m_window.getXcbConnection());
+                }
+                break;
+            }
+        }
+        free(event);
+    }
+}
+
+
+void CWindowXCB::CDispatchThread::init()
+{
+
+}
+
+void CWindowXCB::CDispatchThread::exit()
+{
+}
```

Comment:
btw these methods are wholly optional IIRC (if you don't define them, everything will work anyway

Reply:
`init` isn't but exit is